docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+# CIP: Pipelines, The Registry and Storage
+
+## Status
+
+Current Status: `Under Discussion`
```

Comment:
are we comfortable supporting only python for ops? what is the story for other languages? in particular javascript. 

how would a `Go Client` use pipelines

Reply:
My thought it that:

Pipeline steps are written in python, in the future we can support other execution runtimes but for now, python.
You can write simple pipeline pointers in the client languages that will encode the right step

Reply:
I think that constraint probably makes sense for now (we need some!) 

What do you imagine this would look like in JS then? I guess we could still have types that are chainable... but not runnable locally? hmmm

Reply:
So they could be runnable locally, they would just have to be reimplemented and then if you wanted portability, you'd have to assume they are the same. Or we could make it so that they can never be run locally but I don't love that.

With power comes tradeoffs :(

Reply:
yeah mostly just trying to think through the pros and cons. there will be cons and we will just have to live with them. 

I think for a minimal surface area re-implementation is ok. (like we currently do with embedding functions)

Reply:
> yeah mostly just trying to think through the pros and cons. there will be cons and we will just have to live with them.

Please do. I think we need to hammer this idea into shape, rough edges.

> I think for a minimal surface area re-implementation is ok. (like we currently do with embedding functions)
Yeah I don't see a way around it. But I think 9/10 times you want server-side runs

Reply:
@HammadB @jeffchuber, can this not be only a server-side thing? Then all we need is a DSL or whatever is "natural" in the target language to encode, as suggested above. 

This is how CQL (Chroma Query Language is born)

Reply:
inventing new query languages is a red flag to me! 🚩🚩🚩

Reply:
I think the reason to support client-side is to allow for rapid experimentation.

Reply:
@HammadB, do you think supporting Client and Server side pipelines makes sense?

Since I'm now working on client/server encryption, there is a "world" where encrypting (public key) on the client side and decrypting on the server side (private key) can be helpful.

Reply:
If I understand correctly, the individual operators in the pipeline are stateless, right?
So, there could be a list of operators that you can chain together, given that inputs are compatible with the output after the previous operation in the chain.

Since pipelines are supposed to run server-side, pipelines developed in other languages are not possible. However, what is possible is that each client assembles pipelines based on the operators. I would plan for this from the beginning.

What needs to be created on the client-side in the end to build a pipeline is a manifest file. Most users will likely reuse pipelines, and it makes sense to define these via a manifest in the repository, which also serves as good documentation for the pipeline. I think the idea of using YAML for this is a good one. There doesn't need to be a new query language for that.

In JavaScript, you could create pipelines as a manifest that you send (automatically generated by code, even possible in a pipeline syntax), for example:

```javascript
const collection = client.collection('my-collection');
 pdfHttpStream.pipe(
  chunk(),
  embed(),
  quantize(),
  execute(collection)
);
```
Alternatively, you could pass a pipeline manifest to the collection:

```javascript
const collection = client.collection('my-collection', {pipeline: manifestFile});
collection.add(data);
```

On the server-side, the Python server needs to dynamically build and execute the pipeline based on the manifest.